### PR TITLE
Another refactor on TensorDomain

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- Rename `TensorDomain::{getRootDomain->root, getRFactorDomain->rfactor}`
- `TensorView::{getRootDomain, getRFactorDomain}` remain unchanged

The motivation for this PR is because I feel `domain()->getRootDomain()` is not as good as `domain()->root()`.

I also added an "allocation domain", currently setting the allocation domain is not supported. I added it in this PR just to get a placeholder.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir_internal_nodes.h [hunk 1]
```diff
@@ -1857,20 +1866,34 @@ class TORCH_CUDA_CU_API TensorDomain : public Val {
...
+
+  const std::vector<IterDomain*>& leaf() const {
+    return leaf_domain_;
+  }
+
   // If rfactor domain exists in domain() return it, otherwise return root
   // domain.
   const std::vector<IterDomain*>& getMaybeRFactorDomain() const {
-    return hasRFactor() ? getRFactorDomain() : getRootDomain();
+    return hasRFactor() ? rfactor() : root();
```
Comment (naoyam) [inline csrc/ir_internal_nodes.h:1888]: If we are changing `getRootDomain` to `root`, why not change this to `maybeRfactor`?

File: csrc/ir_internal_nodes.h [hunk 2]
```diff
@@ -1942,10 +1965,12 @@ class TORCH_CUDA_CU_API TensorDomain : public Val {
 
  private:
   const std::vector<IterDomain*> root_domain_;
+  const std::vector<IterDomain*> rfactor_domain_;
+  std::vector<IterDomain*> allocation_domain_;
```
Comment (naoyam) [inline csrc/ir_internal_nodes.h:1969]: This is not used yet, right?

**Review by naoyam — APPROVED** [review]
LGTM

File: csrc/ir_internal_nodes.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/ir_internal_nodes.h:1969]: Correct, it is just a placeholder for now.

File: csrc/ir_internal_nodes.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/ir_internal_nodes.h:1888]: I agree, let me change it.

Comment (zasdfgbnm) [reply csrc/ir_internal_nodes.h:1888]: changed

**Pull request merged**